sheep/recovery.c

```diff
@@ -1151,6 +1158,86 @@ static void screen_object_list(struct recovery_list_work *rlw,
 	xqsort(rlw->oids, rlw->count, obj_cmp);
 }
 
+static int vnode_to_node_idx(struct sd_vnode *vnode, int nr_nodes,
+			     struct sd_node *nodes)
+{
+	for (int i = 0; i < nr_nodes; i++) {
+		if (node_id_cmp(&vnode->node->nid, &nodes[i].nid) == 0)
+			return i;
+	}
+
+	panic("vnode couldn't found in the node array");
+	return -1;		/* never executed */
+}
+
+static bool check_diskfull_possibility(uint32_t epoch, struct vnode_info *vinfo,
+				       int nr_nodes, struct sd_node *nodes)
+{
+	uint64_t **oids_per_node;
+	size_t *nr_oids;
+	uint64_t *required_space_per_node;
+	const struct sd_vnode *vnodes[SD_MAX_COPIES];
+	bool ret = false;
+
+	oids_per_node = xcalloc(nr_nodes, sizeof(uint64_t *));
+	nr_oids = xcalloc(nr_nodes, sizeof(size_t));
+	required_space_per_node = xcalloc(nr_nodes, sizeof(uint64_t));
+
+	for (int i = 0; i < nr_nodes; i++)
+		oids_per_node[i] = fetch_object_list(&nodes[i], epoch,
+						     &nr_oids[i]);
```

Comment:
I think some object IDs can appear in more than one oids_per_node[i]. This will lead to double-counting.